src/lib/serviceRoutines/postQueryContext.cpp

```diff
@@ -241,26 +178,198 @@ std::string postQueryContext
+        }
+        else
+        {
+          cerP->contextElement.contextAttributeVector.push_back(aP);
+        }
+
+        continue;
+      }
+
+      QueryContextRequest* requestP = requestV.lookup(aP->providingApplication, eP);
+
+      if (requestP == NULL)
+      {
+        requestP = new QueryContextRequest(aP->providingApplication, eP, aP->name);
+        requestV.push_back(requestP);
+      }
+      else
+      {
+        requestP->attributeList.push_back_if_absent(aP->name);
+        requestP->entityIdVector.push_back_if_absent(eP);
+      }
+    }
+  }
+
+
+  //
+  // Any local results in localQcrsP?
+  //
+  // If so, localQcrsP must be pushed to the response vector 'responseV'.
+  //
+  if (localQcrsP->contextElementResponseVector.size() != 0)
+  {
+    responseV.push_back(localQcrsP);
   }
+  else
+  {
+    delete localQcrsP;
+  }
+
+
+  //
+  // If no redirectioning is necessary, just return the result
+  //
+  if (requestV.size() == 0)
+  {
+    //
+    // In case the response is empty, fill response with the entity we looked for
+    //
+    if (qcrsP->contextElementResponseVector.size() == 0)
+    {
+      //
+      // FIXME P6: What do we do if we have more than one entity in incoming entityIdVector?
+      //           Push each entityId in a separate contextElementResponse?
+      //           Also, perhaps mongoQueryContext should implement this and not postQueryContext
+      //           This FIXME is related to github issue #588 and (probably) #650.
+      //           Also, optimizing this would be part of issue #768
+      //
+
+
+      //
+      // It's better not to give any details here as we'd have to choose ONE of the entities in the vector ...
+      // std::string details = "Entity id: /" + parseDataP->qcr.res.entityIdVector[0]->id + "/";
+      //
+      qcrsP->errorCode.fill(SccContextElementNotFound);
+    }
+
+    answer = qcrsP->render(ciP, QueryContext, "");
+    return answer;
+  }
+
+
+  //
+  // Now, forward the Query requests, each in a separate thread (to be implemented) and
+  // await all the responses.
+  // Actually, if there is only ONE forward to be done then there is no reason to
+  // do the forward in a separate shell. Better to do it inside the current thread.
+  //
+  // If providingApplication is empty then that part of the query has been performed already, locally.
+  // 
+  //
+  for (unsigned int fIx = 0; fIx < requestV.size(); ++fIx)
+  {
+    if (requestV[fIx]->contextProvider == "")
+    {
+      LM_E(("Internal Error (empty context provider string)"));
+      continue;
+    }
+
+    QueryContextResponse* qP = new QueryContextResponse();
+    qP->errorCode.fill(SccOk);
+    queryForward(ciP, requestV[fIx], qP);
+
+    //
+    // Now, each ContextElementResponse of qP should be tested to see whether there
+    // is already an existing ContextElementResponse in responseV
+    responseV.push_back(qP);
+  }
+
+  answer = responseV.render(ciP, "");
+
+  // requestV.release();
```

Comment:
Anyway, we should try to implement memory management in the right way PR after PR. I mean, not to completely rely on valgrind de-leaking step. At the end, valgrind step is just a tool to improve the code (a good one, by the way) but it could miss some case.